### PR TITLE
Add support for non-optimistic updates (+ API versioning)

### PR DESCRIPTION
**Pull request by khamidou**
**Summary:**
This PR adds a new way to update API objects. Instead of updating API objects before pushing the changes to the backend server (aka optimistic updates), we now push the changes to the backend and wait for the sync engine to pick up the changes. This solves a lot of problems we've had previously with folder and labels, where our datastore and the remote server would get out of sync.

For example, in the case of a folder rename, PUTting changes to `/folders/my_id` will return the title of the original folder. The new name will be returned only once we've carried the folder rename operation. This is a breaking API change, which could be very confusing to our long-time API users.

To keep backward compatibility, I've introduced a very basic API versioning scheme, similar to what Github does with its API. API versions are dates. You can set the API version you'd like to use by setting the "X-Api-Version" header in your request. If you don't set this header, we'll default to the version set for your developer dashboard account (changes forthcoming).

This diff is pretty close to review but I need to take a final pass to make sure everything fits together.

**Test Plan:**
Make a final manual test pass using IMAP and Exchange accounts.

**Reviewers:**
spang, bengotow

### CODE REVIEW

File: inbox/actions/base.py [hunk 1]
```diff
@@ -71,26 +71,30 @@ def create_folder(crispin_client, account_id, category_id):
...
+    new_name = args['new_name']
+    remote_update_folder(crispin_client, account_id, category_id,
+                         old_name, new_name)
 
 
 def delete_folder(crispin_client, account_id, category_id):
     remote_delete_folder(crispin_client, account_id, category_id)
 
 
+def create_label(crispin_client, account_id, category_id):
```
Comment (bengotow) [inline inbox/actions/base.py]: This is a bit of a sidenote, but what's the point of this layer of indirection? It doesn't seem like any of the functions in this file really do anything anymore.

File: inbox/api/ns_api.py [hunk 2]
```diff
@@ -625,20 +641,24 @@ def folder_label_delete_api(public_id):
...
+                folder.deleted_at = deleted_at
 
         schedule_action('delete_folder', category, g.namespace.id,
                         g.db_session)
     else:
-        deleted_at = datetime.utcnow()
-        category.deleted_at = deleted_at
-        for label in category.labels:
-            label.deleted_at = deleted_at
+        if g.api_version == 1:
```
Comment (bengotow) [inline inbox/api/ns_api.py]: This might be cleaner if we translate the API version into API feature flags, and then check those. The comment (`#Optimistically update labels`) wouldn't be as necessary if we could say `if g.api_features.optimistic_updates:`

File: inbox/api/ns_api.py [hunk 3]
```diff
@@ -863,26 +883,43 @@ def event_update_api(public_id):
         # We're going to save this data into a JSON-like TEXT field in the
         # db. With MySQL, this means that the column will be 64k.
         # Drop the latest participants until it fits in the column.
-        while len(json.dumps(cancelled_participants)) > 63000:
+        while len(json.dumps(cancelled_participants)) > 2**15:
```
Comment (bengotow) [inline inbox/api/ns_api.py]: Now 32k, not 63k? Why make the max less than the max size of the column?

File: inbox/api/ns_api.py [hunk 4]
```diff
@@ -863,26 +883,43 @@ def event_update_api(public_id):
...
             cancelled_participants.pop()
 
     # Don't update an event if we don't need to.
     if noop_event_update(event, data):
         return g.encoder.jsonify(event)
 
-    for attr in Event.API_MODIFIABLE_FIELDS:
-        if attr in data:
-            setattr(event, attr, data[attr])
+    if g.api_version == 1:
```
Comment (bengotow) [inline inbox/api/ns_api.py]: Another approach might be to say `should_update_optimistically(g.api_version):`

File: inbox/api/ns_api.py [hunk 5]
```diff
@@ -863,26 +883,43 @@ def event_update_api(public_id):
...
+            if attr in data:
+                setattr(event, attr, data[attr])
 
-    event.sequence_number += 1
-    g.db_session.commit()
+        event.sequence_number += 1
+        g.db_session.commit()
+
+        # Don't sync back updates to autoimported events.
+        # FIXME @karim: document this and forbid it.
```
Comment (bengotow) [inline inbox/api/ns_api.py]: Rather than having this just silently not do anything can we go ahead and return a 400 with a helpful error message? I'd rather people discover the limitations of our API while using it than discover that it sometimes doesn't work for an unknown reason.

File: inbox/api/ns_api.py [hunk 6]
```diff
@@ -863,26 +883,43 @@ def event_update_api(public_id):
...
+        if event.calendar != account.emailed_events_calendar:
+            schedule_action('update_event', event, g.namespace.id, g.db_session,
+                            calendar_uid=event.calendar.uid,
+                            cancelled_participants=cancelled_participants,
+                            notify_participants=notify_participants)
+    else:
+        # This isn't an optimistic update, so we need to store the
+        # updated attributes inside the ActionLog entry.
+        kwargs = dict(calendar_uid=event.calendar.uid,
```
Comment (bengotow) [inline inbox/api/ns_api.py]: There's some code duplication in here that would be nice to avoid. I'm not sure why this input error, for example, is only checked when the update isn't optimistic, and not when it is:

```
if len(json.dumps(kwargs)) > 2**16 - 12:
       raise InputError('Event update too big --- please break it in parts.')
```

Maybe we could generate `kwargs`, run the emailed_events calendar check, and the input length check /before/ we decide what to do?

File: inbox/api/update.py [hunk 7]
```diff
@@ -9,26 +9,28 @@
 # STOPSHIP(emfree): better naming/structure for this module
 
 
-def update_message(message, request_data, db_session):
+def update_message(message, request_data, db_session, optimistic=True):
```
Comment (bengotow) [inline inbox/api/update.py]: Just a preference, but I'd prefer to default optimistic to `None` and `assert(optimistic is not None)` to make sure that this diff isn't overlooking any places this function is called. Defaulting to true could mean we don't find a bug somewhere where the code hasn't been updated and should be passing False.

File: inbox/api/update.py [hunk 8]
```diff
@@ -9,26 +9,28 @@
...
             removed_labels = set(message.categories) - labels
             update_message_labels(message, db_session, added_labels,
-                                  removed_labels)
+                                  removed_labels, optimistic=optimistic)
     else:
         folder = parse_folder(request_data, db_session, message.namespace_id)
         if folder is not None:
-            update_message_folder(message, db_session, folder)
+            update_message_folder(message, db_session, folder,
+                                  optimistic=optimistic)
```
Comment (bengotow) [inline inbox/api/update.py]: (The above would also be nice, because it'd ensure we never forget to do this.) Leaving off `optimistic=optimistic` would cause the wrong behavior here and would be very subtle.

**Comment by bengotow** [comment]
LGTM! Just added a few comments inline.

It'd also be really nice if we could change the API response for non-optimistic PUTs and POSTs. I think it'll be confusing to users that they say unread: false, and we return a Thread with unread: true still. When we implemented this in K2, we made the PUT and POST requests return a task object, which you could query to see if the operation had completed. The task objects were also in the delta stream, so you knew when a particular change had been committed to the provider. It could be a separate PR, but that would allow us to finally change the status indicator in N1 (eg "Applying labels...") so that it reflected when the sync engine finished, not the API request.

**Comment by grinich** [comment]
Do folks have strong feelings on having an integer version number versus a date? The way I've seen Stripe and Clearbit do their gatekeeper versioning is based on a date param that gets sent. More here:

https://stripe.com/docs/upgrades

http://blog.clearbit.com/versioning-apis/

File: inbox/actions/base.py (hunk 1 above)
Comment (khamidou) [inline inbox/actions/base.py]: They let you add behaviour for actions regardless of the provider (e.g: https://github.com/nylas/sync-engine/blob/1d595f2a725b6ee4f2fce8185fc7efa7a07ecac3/inbox/actions/base.py#L121)

File: inbox/api/ns_api.py (hunk 5 above)
Comment (khamidou) [inline inbox/api/ns_api.py]: Good point.

File: inbox/api/ns_api.py (hunk 6 above)
Comment (khamidou) [inline inbox/api/ns_api.py]: Basically there's two flows here:
1. Optimistic updates. We update the event object in-place and save it. The ActionLog worker will use this event object to update the remote backend.
2. Non-optimistic updates. We save the changes we've received to the ActionLog's extra field (a JSON column). Because this column is a TEXT column, it's limited to 64k chars, so we need to check there's enough space for everything.

File: inbox/api/ns_api.py (hunk 2 above)
Comment (khamidou) [inline inbox/api/ns_api.py]: Good point, will add this.

File: inbox/api/ns_api.py (hunk 3 above)
Comment (khamidou) [inline inbox/api/ns_api.py]: Hmm, looks like a mistake. Fixing.

File: inbox/api/ns_api.py (hunk 5 above)
Comment (khamidou) [inline inbox/api/ns_api.py]: Good point, fixing.

File: inbox/api/ns_api.py (hunk 4 above)
Comment (khamidou) [inline inbox/api/ns_api.py]: Fixed!

**Comment by khamidou** [comment]
I have no preferences about it. The latest version of this branch uses a date field instead of an integer version.

**Comment by grinich** [comment]
Cool, I personally like the date field since it's more explicitly about when the change happened. We should probably also have an API changelog recorded somewhere.

File: inbox/api/srv.py [hunk 9]
```diff
@@ -130,4 +130,5 @@ def logout():
 
 
 app.register_blueprint(ns_api)
+app.register_blueprint(ns_api, url_prefix='/v2')
```
Comment (spang) [inline inbox/api/srv.py]: ...what is this for? cruft before we decided to go with gating rather than changing the URL?

File: tests/api/test_events.py [hunk 10]
```diff
@@ -152,6 +153,40 @@ def test_api_update_title(db, api_client, calendar, default_account):
...
+@pytest.mark.only
+def test_api_pessimistic_update(db, api_client, calendar, default_account):
+    e_data = {
+        'title': '',
+        'calendar_id': calendar.public_id,
+        'when': {'time': 1407542195},
+    }
+
+    e_resp = api_client.post_data('/events', e_data,
+                                  headers={ "X-API-Version": API_VERSIONS[1] })
```
Comment (spang) [inline tests/api/test_events.py]: Is this the same header that other APIs which do this use?

File: inbox/transactions/actions.py [hunk 11]
```diff
@@ -222,9 +222,9 @@ def _process_log(self):
                 else:
                     namespaces_to_process = random.sample(namespace_ids,
                                                           NUM_PARALLEL_ACCOUNTS)
-                self.log.info('Syncback namespace_ids count', shard_id=key,
-                              process=self.process_number,
-                              num_namespace_ids=len(namespace_ids))
+                #self.log.info('Syncback namespace_ids count', shard_id=key,
```
Comment (spang) [inline inbox/transactions/actions.py]: Is this related? Not sure what it has to do with this patch

File: inbox/transactions/actions.py [hunk 12]
```diff
@@ -239,10 +239,10 @@ def _process_log(self):
                         self.task_queue.put(task)
 
         after = datetime.utcnow()
-        self.log.info('Syncback completed one iteration',
-                      process=self.process_number,
-                      duration=(after - before).total_seconds(),
-                      idle_workers=self.num_idle_workers)
+        #self.log.info('Syncback completed one iteration',
```
Comment (spang) [inline inbox/transactions/actions.py]: or this

File: inbox/api/ns_api.py [hunk 13]
```diff
@@ -300,7 +319,12 @@ def thread_api_update(public_id):
     data = request.get_json(force=True)
     if not isinstance(data, dict):
         raise InputError('Invalid request body')
-    update_thread(thread, data, g.db_session)
+
+    if g.api_features.optimistic_updates:
+        update_thread(thread, data, g.db_session, True)
```
Comment (spang) [inline inbox/api/ns_api.py]: any reason we're not just passing `g.api_features.optimistic_updates` in as the argument to `update_thread()`. Seems kind of weird to put a conditional on a boolean and then explicitly pass in the exact same boolean when the same thing could be accomplished in a single line of code

**Comment by spang** [comment]
This in general LGTM. Can you explain the folder and calendar pieces of this? They're pretty complex and I don't quite understand how they work.

File: inbox/transactions/actions.py (hunk 11 above)
Comment (khamidou) [inline inbox/transactions/actions.py]: Oh I commented those because they spam the console and make it super annoying to debug API requests.

File: tests/api/test_events.py (hunk 10 above)
Comment (khamidou) [inline tests/api/test_events.py]: No. Stripe uses `Stripe-Version`. Clearbit uses `Api-Version`. Apparently, it's not recommended anymore to append `X-` to HTTP headers, so I'll just go with `Api-Version`.

File: inbox/api/ns_api.py (hunk 13 above)
Comment (khamidou) [inline inbox/api/ns_api.py]: Good point.

File: inbox/api/srv.py (hunk 9 above)
Comment (khamidou) [inline inbox/api/srv.py]: Yup. Removed it.

**Comment by khamidou** [comment]
Sure. 

Concerning folders, the main issue we're having is that we're slow to detect folder renames/deletes (see: https://phab.nylas.com/T6988), so to save time, we simply update the folder once we've got confirmation from the IMAP server that the rename went correctly.

We have a similar problem for event updates --- because of the way ActiveSync works, we _have_ to update the event once the server acks our request. To make this work, we store the updated fields in the `ActionLog` entry's `extra_args` fields (it's a JSON TEXT field, which is why there's some logic to make sure updates aren't bigger than 64k). Note that we're doing the same thing for Google Calendar --- we could have waited for the server to notify us but I preferred to share the same logic for event updates.

I've updated the diff to address your other comments:
- you can now set the API version by using the `Api-Version` HTTP header
- cleaned up the cruft

**Comment by spang** [comment]
LGTM! Let's take this to staging! :)
